api/src/license_operator.rs

```diff
@@ -0,0 +1,207 @@
+use std::borrow::Borrow;
+
+use log::info;
+use rand::rngs::ThreadRng;
+use rand::Rng;
+use sha3::{
+    digest::{ExtendableOutput, Update, XofReader},
+    Shake256,
+};
+
+use crate::adler32::adler32_checksum;
+use crate::license_key::LicenseKeyStatus;
+use crate::license_magic::LicenseMagic;
+use crate::license_serializer::{DefaultLicenseKeySerializer, LicenseKeySerializer};
+use crate::LicenseKey;
+
+pub struct LicenseOperator {
+    rng: ThreadRng,
+
+    magic: LicenseMagic,
+    key_size: usize,
+    checksum_init: (u32, u32),
+
+    serializer: Box<dyn LicenseKeySerializer>,
+
+    blacklist: Vec<Vec<u8>>,
+    byte_check: Vec<(usize, LicenseMagic)>,
+}
+
+impl LicenseOperator {
+    #[inline(always)]
+    pub fn new(
+        magic: LicenseMagic,
+        key_size: usize,
+        checksum_init: (u32, u32),
+        serializer: Box<dyn LicenseKeySerializer>,
+        blacklist: Vec<Vec<u8>>,
+        byte_check: Vec<(usize, LicenseMagic)>,
+    ) -> LicenseOperator {
+        LicenseOperator {
+            rng: rand::thread_rng(),
+            magic,
+            key_size,
+            checksum_init,
+            serializer,
+            blacklist,
+            byte_check,
+        }
+    }
+
+    #[inline(always)]
+    pub fn default() -> LicenseOperator {
+        LicenseOperator {
+            rng: rand::thread_rng(),
+            magic: LicenseMagic::default(),
+            key_size: 16,
+            checksum_init: (0xFFAA, 0xAAFF),
+            serializer: Box::new(DefaultLicenseKeySerializer {}),
+            blacklist: Vec::new(),
+            byte_check: Vec::new(),
+        }
+    }
+
+    #[inline(always)]
+    pub fn randomize_magic(&mut self, magic_size: usize, magic_count: usize) -> &Self {
+        for _x in 0..magic_size {
+            let mut magic: Vec<u8> = Vec::new();
+
+            info!("Randomized magic:");
+            for y in 0..magic_count {
+                let random_value = self.rng.gen();
+                magic.push(random_value);
+                info!("\t {} - {}", y, random_value);
+            }
+
+            self.magic.push(magic);
+        }
+
+        self
+    }
+
+    #[inline(always)]
+    pub fn get_serialized_key(&self, license_key: &LicenseKey) -> String {
+        self.serializer.serialize_key(&license_key.serialized_key)
+    }
+
+    #[inline(always)]
+    pub fn add_seed_to_blacklist(&mut self, seed: &[u8]) -> &Self {
+        self.blacklist.push(seed.to_vec());
+
+        self
+    }
+
+    #[inline(always)]
+    pub fn generate_license_key(&self, seed: &[u8]) -> LicenseKey {
+        // Validate user parameters
+        // Minimal 8 seed size, USER_PAYLOAD payload size and 4 checksum size
+        let license_key_required_size: usize = 8 + self.magic.payload_size() + 4;
+        if self.key_size <= license_key_required_size {
+            info!(
```

Comment:
I rather used `color-eyre` cargo for `Report` than `Box<dyn Error>`.

Reply:
Yeah maybe, however it should be as easily installable lib, that is packed without many third party deps. Let the user handle errors how he wants, you know what I mean? So I would much rather create my custom errors and let the user wrap them. Or create multiple versions that can be disabled using feature flags. I am not really fond of dependencies, that may someday ceased to be supported, especially the one, that are not bringing any real value, but visuals.

However, this is your repo, so the decision have to made by you. 🙂

Reply:
Well, that's make sense for me to use simple std error instead of eyre...

So what I used was `simple-error` which have macro for return `std::error::Error` for me easily with `bail!()` and used eyre only in examples(APP) for better output when something go wrong.